File: Wanda/ROS/src/walking_gait/nodes/twist_lowpass_filter_node.py
#!/usr/bin/env python3
#=====================
import copy
import rospy

# Standard ROS message type
from geometry_msgs.msg import TwistStamped

# Custom libraries
from topic_messages import topics

top = topics.topics()

#==============================================================================
class twist_lowpass_filter_node():
    def __init__(self):
        rospy.init_node( "twist_lowpass_filter_node", anonymous=True )

        self.twist_old = TwistStamped()
        self.twist_old.twist = self.init_twist( self.twist_old.twist )

        self.twist_new = TwistStamped()
        self.twist_new.twist = self.init_twist( self.twist_new.twist )

        self.period = 1.0/5.0 # Filter frequency

        self.pub = rospy.Publisher( top.walking_twist_filtered, TwistStamped, queue_size=1 )

        rospy.Subscriber( top.walking_twist_commanded, TwistStamped, self.filter_twist )

    #======================================================
    def filter_twist( self, twist_in ):

        self.twist_new = copy.deepcopy( twist_in )

        # Filter the components of the twist
        self.twist_new.twist.linear.x = self.low_pass_filter(
                                            self.twist_new.twist.linear.x,
                                            self.twist_old.twist.linear.x )
        self.twist_new.twist.linear.y = self.low_pass_filter(
                                            self.twist_new.twist.linear.y,
                                            self.twist_old.twist.linear.y )
        self.twist_new.twist.linear.z = self.low_pass_filter(
                                            self.twist_new.twist.linear.z,
                                            self.twist_old.twist.linear.z )

        self.twist_new.twist.angular.x = self.low_pass_filter(
                                            self.twist_new.twist.angular.x,
                                            self.twist_old.twist.angular.x )
        self.twist_new.twist.angular.y = self.low_pass_filter(
                                            self.twist_new.twist.angular.y,
                                            self.twist_old.twist.angular.y )
        self.twist_new.twist.angular.z = self.low_pass_filter(
                                            self.twist_new.twist.angular.z,
                                            self.twist_old.twist.angular.z )

        # Send out the filtered twist
        self.pub.publish( self.twist_new )

        # Make sure to save this new command for next time
        self.twist_old = copy.deepcopy( self.twist_new )

    #======================================================
    def low_pass_filter( self, state_current, state_old, period=None ):
        if period == None:
            # Use the default filter frequency if none is specified
            period = self.period

        state_new = period * state_old + (1-period) * state_current

        return state_new

    #======================================================
    def init_twist( self, twist ):
        twist.linear.x = 0.0
        twist.linear.y = 0.0
        twist.linear.z = 0.0
        twist.angular.x = 0.0
        twist.angular.y = 0.0
        twist.angular.z = 0.0
        return twist

#==============================================================================
if __name__ == "__main__":
    try:
        tf = twist_lowpass_filter_node()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: Wanda/ROS/src/walking_gait/src/twist_lowpass_filter_node.cpp
#include <ros/ros.h>

// Custom classes
#include "system_globals/frames.h"
#include "walking_gait/twist_lowpass_filter.h"

//===============================================
// Constructor
twist_lowpass_filter::twist_lowpass_filter( ros::NodeHandle* nh_in ):
nh( *nh_in ),
filter_period( 1.0 / 3.0 )
{
    // Initialize my variables
    twist_commanded.twist = init_twist();
    twist_filtered.twist = init_twist();
    // Make sure the output twist has a frame, at least to start with
    // This will be overwritten with the frame in the commanded twist
    // once that starts coming i n
    Frames frames( nh );
    twist_commanded.header.frame_id = frames.ground();
    twist_filtered.header.frame_id = frames.ground();

    process_user_inputs();

    // Initialize ROS-stuff
    pub = nh.advertise<geometry_msgs::TwistStamped>( published_topic_name, 1 );
    sub = nh.subscribe( subscribed_topic_name,
                        10,
                        &twist_lowpass_filter::get_twist_commanded,
                        this );
}

//===============================================
void twist_lowpass_filter::process_user_inputs()
{
    std::string node_name = ros::this_node::getName();

    Topics top( nh );
    nh.param<std::string>( node_name + "/topic_in",
                           subscribed_topic_name,
                           top.walking_twist_commanded() );

    nh.param<std::string>( node_name + "/topic_out",
                           published_topic_name,
                           top.walking_twist_filtered() );
}

//===============================================
void twist_lowpass_filter::get_twist_commanded(
    const geometry_msgs::TwistStamped& twist
)
{
    twist_commanded = twist;
}

//===============================================
void twist_lowpass_filter::filter_twist()
{

    ros::Rate loop_rate( 30 );

    while ( nh.ok() ){

        // Filter each of the filter states
        twist_filtered.twist.linear.x = low_pass_filter( twist_filtered.twist.linear.x,
                                                         twist_commanded.twist.linear.x,
                                                         filter_period );

        twist_filtered.twist.linear.y = low_pass_filter( twist_filtered.twist.linear.y,
                                                         twist_commanded.twist.linear.y,
                                                         filter_period );

        twist_filtered.twist.linear.z = low_pass_filter( twist_filtered.twist.linear.z,
                                                         twist_commanded.twist.linear.z,
                                                         filter_period );

        twist_filtered.twist.angular.x = low_pass_filter( twist_filtered.twist.angular.x,
                                                          twist_commanded.twist.angular.x,
                                                          filter_period );

        twist_filtered.twist.angular.y = low_pass_filter( twist_filtered.twist.angular.y,
                                                          twist_commanded.twist.angular.y,
                                                          filter_period );

        twist_filtered.twist.angular.z = low_pass_filter( twist_filtered.twist.angular.z,
                                                          twist_commanded.twist.angular.z,
                                                          filter_period );

        // Grab header stuff from the twist
        twist_filtered.header.frame_id = twist_commanded.header.frame_id;
        twist_filtered.header.stamp = ros::Time::now();

        // Send out the data
        pub.publish( twist_filtered );

        ros::spinOnce();
        loop_rate.sleep();
    }
}

//===============================================
float twist_lowpass_filter::low_pass_filter(
    float state_filter,
    float state_desired,
    float period
)
{
    float state_new;
    state_new = state_filter + period * ( state_desired - state_filter );
    return state_new;
}

//===============================================
geometry_msgs::Twist twist_lowpass_filter::init_twist()
{
    geometry_msgs::Twist twist;
    twist.linear.x = 0.0;
    twist.linear.y = 0.0;
    twist.linear.z = 0.0;
    twist.angular.x = 0.0;
    twist.angular.y = 0.0;
    twist.angular.z = 0.0;
    return twist;
}

//=============================================================================
int main( int argc, char** argv){

    ros::init(argc, argv, "twist_lowpass_filter_node");
    ros::NodeHandle node_handler;

    twist_lowpass_filter twist_lowpass_filter( &node_handler );
    twist_lowpass_filter.filter_twist();
    ros::spin();

    return 0;
}

